controllers/configuration/backend/backend.go

```diff
@@ -0,0 +1,316 @@
+	"github.com/hashicorp/hcl/v2/hclparse"
+	"github.com/hashicorp/hcl/v2/hclwrite"
+	"github.com/oam-dev/terraform-controller/api/v1beta2"
+	"github.com/pkg/errors"
+	"github.com/zclconf/go-cty/cty"
+	"github.com/zclconf/go-cty/cty/gocty"
+	"k8s.io/klog/v2"
+)
+
+var backendSecretMap = map[v1beta2.BackendType]map[string]string{
+	"azurerm": {
+		"client_certificate_path": "ClientCertificateSecret",
+	},
+	"consul": {
+		"ca_file":   "CAFileSecret",
+		"cert_file": "CertFileSecret",
+		"key_file":  "KeyFileSecret",
+	},
+	"etcdv3": {
+		"cacert_path": "CacertSecret",
+		"cert_path":   "CertSecret",
+		"key_path":    "KeySecret",
+	},
+	"gcs": {
+		"credentials": "CredentialsSecret",
+	},
+	"kubernetes": {
+		"config_path":  "ConfigSecret",
+		"config_paths": "ConfigSecrets",
+	},
+	"oss": {
+		"shared_credentials_file": "SharedCredentialsSecret",
+	},
+	"s3": {
+		"shared_credentials_file": "SharedCredentialsSecret",
+	},
+	"swift": {
+		"cacert_file": "CacertSecret",
+		"cert":        "CertSecret",
+		"key":         "KeySecret",
+	},
+}
+
+var backendTF = `
+terraform {
+  backend "kubernetes" {
+    secret_suffix     = "{{.SecretSuffix}}"
+    in_cluster_config = {{.InClusterConfig}}
+    namespace         = "{{.Namespace}}"
+  }
+}
+`
+
+// Conf contains all the backend information in the meta
+type Conf struct {
+	BackendType v1beta2.BackendType
+	HCL         string
+	// UseCustom indicates whether to use custom kubernetes backend
+	UseCustom bool
+	// Secrets describes which secret and which key in the secret should be mounted to the executor pod
+	Secrets []*ConfSecretSelector
+}
+
+// ConfSecretSelector describes which secret should be mounted to the executor pod
+type ConfSecretSelector struct {
+	// Name is the name of the secret which will be mounted to a pod when running the job
+	Name string
+	// Path is a relative path, indicates which path the secret should be mounted to
+	Path string
+	// Keys are the keys that contains the file content in secret
+	Keys []string
+}
+
+type backendVars struct {
+	SecretSuffix    string
+	InClusterConfig bool
+	Namespace       string
+}
+
+// renderTemplate renders Backend template
+func renderTemplate(backend *v1beta2.Backend, namespace string) (string, error) {
+	tmpl, err := template.New("backend").Funcs(template.FuncMap(sprig.FuncMap())).Parse(backendTF)
+	if err != nil {
+		return "", err
+	}
+
+	templateVars := backendVars{
+		SecretSuffix:    backend.SecretSuffix,
+		InClusterConfig: backend.InClusterConfig,
+		Namespace:       namespace,
+	}
+	var wr bytes.Buffer
+	err = tmpl.Execute(&wr, templateVars)
+	if err != nil {
+		return "", err
+	}
+	return wr.String(), nil
+}
+
+// ParseConfigurationBackend parses backend Conf from the v1beta2.Configuration
+func ParseConfigurationBackend(configuration *v1beta2.Configuration, terraformBackendNamespace string) (*Conf, error) {
+	backend := configuration.Spec.Backend
+
+	switch {
+
+	case backend != nil && len(backend.Inline) > 0:
```

Comment:
What if `Inline` is not empty and BackType is not empty?

Reply:
Then we will use the `inline` and ignore the `backendType`.